File: 1544-count-good-nodes-in-binary-tree/count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int goodNodes(TreeNode root) {
        return countGoodNodes(root, Integer.MIN_VALUE);
    }
    
    private int countGoodNodes(TreeNode node, int maxSoFar) {
        if (node == null) {
            return 0;
        }
        
        int count = (node.val >= maxSoFar) ? 1 : 0;
        int newMax = Math.max(maxSoFar, node.val);
        
        count += countGoodNodes(node.left, newMax);
        count += countGoodNodes(node.right, newMax);
        
        return count;
    }
}

File: 1544-count-good-nodes-in-binary-tree/count-good-nodes-in-binary-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

int countGoodNodes(struct TreeNode* root, int maxSoFar) {
    if (root == NULL) {
        return 0;
    }
    
    int count = (root->val >= maxSoFar) ? 1 : 0;
    int newMax = (root->val > maxSoFar) ? root->val : maxSoFar;
    
    count += countGoodNodes(root->left, newMax);
    count += countGoodNodes(root->right, newMax);
    
    return count;
}

int goodNodes(struct TreeNode* root) {
    return countGoodNodes(root, INT_MIN);
}